### DISCUSSION TITLE
event hook after request, but before serialization?

### DISCUSSION THREAD

**Discussion by cfal** [fastapi/fastapi#8933] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

I'd like to be able to check the `BaseModel` objects returned by a route before they are serialized. what would be the best way to do that? i couldn't find any obvious callbacks where i could get the returned objects passed in before serialization. is this possible with middleware?

**Comment by cfal** [comment, marked as answer]
i ended up creating a custom APIRouter that overrides `add_api_route` and wraps the endpoint with a function that copies the endpoint's function signature. still interested if there are other options though :)

**Comment by cheesycod** [comment]
Is it not possible to use response_model

**Comment by cfal** [comment]
> Is it not possible to use response_model

so one thing I forgot to mention is that.. I have to check and validate the returned `BaseModel`'s using the current user, ie. I have a `Depends(get_user)` parameter in the endpoint, that I would like to use. I tried with response_model, and although in fastapi's `routing.py` class I did see that I could override `BaseModel.json()` and do some checking there, I couldn't find an easy way to reuse my injected dependency. or is there a way?
